app/scripts/ui.js

```diff
@@ -32,24 +76,26 @@ async function start() {
 
   // setup stream to background
   const extensionPort = browser.runtime.connect({ name: windowType });
+
+  extensionPort.onDisconnect.addListener((_port) => {
+    const container = document.getElementById('app-content');
+    displayCriticalError(
+      container,
+      new Error('Connection lost to background script'),
+      'backgroundPortClosedError',
+    );
+  });
+
   const connectionStream = new PortStream(extensionPort);
 
   const activeTab = await queryCurrentActiveTab(windowType);
   initializeUiWithTab(activeTab);
 
-  function displayCriticalError(container, err) {
-    container.innerHTML =
-      '<div class="critical-error">The MetaMask app failed to load: please open and close MetaMask again to restart.</div>';
-    container.style.height = '80px';
-    log.error(err.stack);
-    throw err;
-  }
-
   function initializeUiWithTab(tab) {
     const container = document.getElementById('app-content');
     initializeUi(tab, container, connectionStream, (err, store) => {
       if (err) {
-        displayCriticalError(container, err);
+        displayCriticalError(container, err, 'failedToLoadMessage');
         return;
       }
```

Comment:
At times frontend does not throw error but there is white screen for long, if RPC connection can not be established between background and UI. I am not sure if these changes take care of that scenario also.

Reply:
Showing a loading screen initially instead of a blank screen should address this problem. Presumably the connection attempt will eventually timeout. If not, we can add a timeout.

Reply:
https://github.com/MetaMask/metamask-extension/pull/14461#issuecomment-1105069726